narwhals/_expression_parsing.py

```diff
@@ -342,3 +343,44 @@ def operation_is_order_dependent(*args: IntoExpr | Any) -> bool:
     # it means that it was a scalar (e.g. nw.col('a') + 1) or a column name,
     # neither of which is order-dependent, so we default to `False`.
     return any(getattr(x, "_is_order_dependent", False) for x in args)
+
+
+def operation_changes_length(*args: IntoExpr | Any) -> bool:
+    """Track whether operation changes length.
+
+    n-ary operations between expressions which change length are not
+    allowed. This is because the output might be non-relational. For
+    example:
+        df = pl.LazyFrame({'a': [1,2,None], 'b': [4,None,6]})
+        df.select(pl.col('a', 'b').drop_nulls())
+    Polars does allow this, but in the result we end up with the
+    tuple (2, 6) which wasn't part of the original data.
+
+    Rules are:
+        - in an n-ary operation, if any one of them changes length, then
+          it must be the only expression present
+        - in a comparison between a changes-length expression and a
+          scalar, the output changes length
+    """
+    from narwhals.expr import Expr
+
+    n_exprs = len([x for x in args if isinstance(x, Expr)])
+    changes_length = any(
+        isinstance(x, Expr) and x._changes_length and not x._aggregates for x in args
+    )
+    if n_exprs > 1 and changes_length:
+        msg = (
+            "Found multiple expressions at least one of which changes length. You can only use one "
+            "length-changing expression at a time, unless it is followed by an aggregation."
+        )
```

Comment:
I find this message a bit confusing. I imagine you mean that a length changing expression might be used only if it's the only expression.

Maybe:

```suggestion
        msg = (
            "Found multiple expressions at least one of which changes length. Any length-changing expression "
            "can be used in isolation, unless it is followed by an aggregation."
        )
```